quantum/pointing_device.c

```diff
@@ -54,67 +145,283 @@ __attribute__((weak)) void pointing_device_init(void) {
+    local_mouse_report.buttons     = local_mouse_report.buttons | mousekey_report.buttons;
 #endif
     pointing_device_send();
 }
 
-report_mouse_t pointing_device_get_report(void) { return mouseReport; }
+/**
+ * @brief Gets current mouse report used by pointing device task
+ *
+ * @return report_mouse_t
+ */
+report_mouse_t pointing_device_get_report(void) { return local_mouse_report; }
+
+/**
+ * @brief Sets mouse report used be pointing device task
+ *
+ * @param[in] new_mouse_report
+ */
+void pointing_device_set_report(report_mouse_t new_mouse_report) { local_mouse_report = new_mouse_report; }
+
+/**
+ * @brief Gets current pointing device CPI if supported
+ *
+ * Gets current cpi from pointing device driver if supported and returns it as uint16_t
+ *
+ * @return cpi value as uint16_t
+ */
+uint16_t pointing_device_get_cpi(void) {
+#if defined(SPLIT_POINTING_ENABLE)
+    return POINTING_DEVICE_THIS_SIDE ? pointing_device_driver.get_cpi() : shared_cpi;
+#else
+    return pointing_device_driver.get_cpi();
+#endif
+}
+
+/**
+ * @brief Set pointing device CPI if supported
+ *
+ * Takes a uint16_t value to set pointing device cpi if supported by driver.
+ *
+ * @param[in] cpi uint16_t value.
+ */
+void pointing_device_set_cpi(uint16_t cpi) {
+#if defined(SPLIT_POINTING_ENABLE)
+    if (POINTING_DEVICE_THIS_SIDE) {
+        pointing_device_driver.set_cpi(cpi);
+    } else {
+        shared_cpi = cpi;
+    }
+#else
+    pointing_device_driver.set_cpi(cpi);
+#endif
+}
+
+#if defined(SPLIT_POINTING_ENABLE) && defined(POINTING_DEVICE_COMBINED)
+/**
+ * @brief Set pointing device CPI if supported
+ *
+ * Takes a bool and uint16_t and allows setting cpi for a single side when using 2 pointing devices with a split keyboard.
+ *
+ * NOTE: Only available when using SPLIT_POINTING_ENABLE and POINTING_DEVICE_COMBINED
+ *
+ * @param[in] left true = left, false = right.
+ * @param[in] cpi uint16_t value.
+ */
+void pointing_device_set_cpi_on_side(bool left, uint16_t cpi) {
+    bool local = (is_keyboard_left() & left) ? true : false;
+    if (local) {
+        pointing_device_driver.set_cpi(cpi);
+    } else {
+        shared_cpi = cpi;
+    }
+}
+
+/**
+ * @brief combines 2 mouse reports and returns 2
+ *
+ * Takes 2 report_mouse_t structs, performs an inclusive or ignoring report_id then returns the resulting report_mouse_t struct.
+ *
+ * NOTE: Only available when using SPLIT_POINTING_ENABLE and POINTING_DEVICE_COMBINED
+ *
+ * @param[in] left_report left report_mouse_t
+ * @param[in] right_report right report_mouse_t
+ * @return combined report_mouse_t of left_report and right_report
+ */
+report_mouse_t pointing_device_combine_reports(report_mouse_t left_report, report_mouse_t right_report) {
+    left_report.x |= right_report.x;
+    left_report.y |= right_report.y;
+    left_report.h |= right_report.h;
+    left_report.v |= right_report.v;
+    left_report.buttons |= right_report.buttons;
+    return left_report;
+}
```

Comment:
```suggestion
report_mouse_t pointing_device_combine_reports(report_mouse_t left_report, report_mouse_t right_report) {
    left_report.x = CLAMP(left_report.x + right_report.x, INT8_MIN, INT8_MAX);
    left_report.y = CLAMP(left_report.y + right_report.y, INT8_MIN, INT8_MAX);
    left_report.h = CLAMP(left_report.h + right_report.h, INT8_MIN, INT8_MAX);
    left_report.v = CLAMP(left_report.v + right_report.v, INT8_MIN, INT8_MAX);
    left_report.buttons |= right_report.buttons;
    return left_report;
}
```
OR'ing together numerics is a debatable way of combining values -- if we have one side positive and one negative, then the result will always be negative anyway.
The other consideration is that if there's any sort of "noise" that results in one side returning, say, `-1`, then the bitwise representation of the `int8_t` will be `0xFF` and any value on the other report will effectively be ignored.

Suggest the above code where we add the values and then clamp the result to the max range of the `int8_t`.

Reply:
Yeah definitely missed this, I've implemented a clamp function. Thanks.